comms/src/outbound_message_service/outbound_message.rs

```diff
@@ -0,0 +1,110 @@
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    connection::message::{Frame, MessageError},
+    peer_manager::node_id::NodeId,
+};
+use chrono::prelude::*;
+use rmp_serde;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::convert::TryFrom;
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct OutboundMessage<T> {
+    pub destination_node_id: NodeId,
+    pub retry_count: u32,
+    pub creation_timestamp: DateTime<Utc>,
+    pub last_retry_timestamp: Option<DateTime<Utc>>,
+    pub message_envelope: T,
+}
+
+impl<T: Serialize + DeserializeOwned> OutboundMessage<T> {
+    /// Create a new OutboundMessage from the destination_node_id and message_envelope
+    pub fn new(destination_node_id: NodeId, message_envelope: T) -> OutboundMessage<T> {
+        OutboundMessage {
+            destination_node_id,
+            retry_count: 0,
+            creation_timestamp: Utc::now(),
+            last_retry_timestamp: None,
+            message_envelope,
+        }
+    }
+
+    /// Serialize an OutboundMessage into a single frame
+    pub fn to_frame(&self) -> Result<Frame, MessageError> {
+        let mut buf: Vec<u8> = Vec::new();
+        match self.serialize(&mut rmp_serde::Serializer::new(&mut buf)) {
+            Ok(_) => Ok(buf.to_vec()),
+            Err(_) => Err(MessageError::SerializeFailed),
+        }
+    }
+
+    /// Update the retry count and retry timestamp after a transmission attempt
+    pub fn transmission_attempt(&mut self) {
+        self.retry_count += 1;
+        self.last_retry_timestamp = Some(Utc::now());
+    }
+}
+
+impl<T: Serialize + DeserializeOwned> TryFrom<Frame> for OutboundMessage<T> {
+    type Error = MessageError;
+
+    /// Construct an OutboundMessage from a Frame
+    fn try_from(frame: Frame) -> Result<Self, Self::Error> {
+        let mut de = rmp_serde::Deserializer::new(frame.as_slice());
+        match Deserialize::deserialize(&mut de) {
+            Ok(outbound_message) => Ok(outbound_message),
+            Err(_) => Err(MessageError::DeserializeFailed),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use tari_crypto::{
+        keys::{PublicKey, SecretKey},
+        ristretto::{RistrettoPublicKey, RistrettoSecretKey},
+    };
+
+    #[test]
+    fn test_outbound_message() {
+        let mut rng = rand::OsRng::new().unwrap();
+        let sk = RistrettoSecretKey::random(&mut rng);
+        let pk = RistrettoPublicKey::from_secret_key(&sk);
```

Comment:
fyi, There is a PublicKey::random_pair() method

Reply:
That looks cleaner, change applied